storage/database/db_manager.go

```diff
@@ -1448,6 +1453,109 @@ func (dbm *databaseManager) DeleteBlock(hash common.Hash, number uint64) {
 	dbm.cm.deleteBlockCache(hash)
 }
 
+const badBlockToKeep = 100
+
+type badBlock struct {
+	Header *types.Header
+	Body   *types.Body
+}
+
+// badBlockList implements the sort interface to allow sorting a list of
+// bad blocks by their number in the reverse order.
+type badBlockList []*badBlock
+
+func (s badBlockList) Len() int { return len(s) }
+func (s badBlockList) Less(i, j int) bool {
+	return s[i].Header.Number.Uint64() < s[j].Header.Number.Uint64()
+}
+func (s badBlockList) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+
+// ReadBadBlock retrieves the bad block with the corresponding block hash.
+func (dbm *databaseManager) ReadBadBlock(hash common.Hash) *types.Block {
+	db := dbm.getDatabase(MiscDB)
+	blob, err := db.Get(badBlockKey)
+	if err != nil {
+		return nil
+	}
+	var badBlocks badBlockList
+	if err := rlp.DecodeBytes(blob, &badBlocks); err != nil {
+		return nil
+	}
+	for _, bad := range badBlocks {
+		if bad.Header.Hash() == hash {
+			return types.NewBlockWithHeader(bad.Header).WithBody(bad.Body.Transactions)
+		}
+	}
+	return nil
+}
+
+// ReadAllBadBlocks retrieves all the bad blocks in the database.
+// All returned blocks are sorted in reverse order by number.
+func (dbm *databaseManager) ReadAllBadBlocks() ([]*types.Block, error) {
+	var badBlocks badBlockList
+	db := dbm.getDatabase(MiscDB)
+	blob, err := db.Get(badBlockKey)
+	if err != nil {
+		return nil, err
+	}
+	if err := rlp.DecodeBytes(blob, &badBlocks); err != nil {
+		return nil, err
+	}
+	blocks := make([]*types.Block, len(badBlocks))
+	for i, bad := range badBlocks {
+		blocks[i] = types.NewBlockWithHeader(bad.Header).WithBody(bad.Body.Transactions)
+	}
+	return blocks, nil
+}
+
+// WriteBadBlock serializes the bad block into the database. If the cumulated
+// bad blocks exceed the capacity, the oldest will be dropped.
+func (dbm *databaseManager) WriteBadBlock(block *types.Block) {
```

Comment:
The reason to read all badblocks is to store badblocks in order. I agree this operation itself is costly high, but since max length is 100, it's not a big deal.